File: apps/angular-app/src/app/features/funds/domain/repositories/fund.repository.ts
import { Observable } from 'rxjs';
import { Result } from '../../../../core/errors/either';
import { Fund } from '../entities/fund.entity';

export abstract class FundRepository {
  abstract getFunds(): Observable<Result<Fund[]>>;
  abstract subscribeTo(fundId: number, amount: number, notification: string): Observable<Result<void>>;
}

File: apps/flutter-app/lib/features/funds/domain/repositories/fund_repository.dart
import 'package:dartz/dartz.dart';
import 'package:fondos_btg/core/error/failures.dart';
import 'package:fondos_btg/features/funds/domain/entities/fund.dart';

abstract class FundRepository {
  /// Returns all available funds.
  Future<Either<Failure, List<Fund>>> getFunds();

  /// Subscribes a user to a fund.
  /// Returns the updated balance on success.
  Future<Either<Failure, double>> subscribeToFund({
    required int fundId,
    required double amount,
    required String notificationMethod,
  });
}
